src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -219,17 +219,31 @@
   "Initializes a 'worker' which will listen for lifecycle-related tasks and perform
   them on a background thread, to ensure that we aren't executing multiple lifecycle
   tasks simultaneously."
-  [lifecycle-chan :- (schema/protocol async-prot/Channel)
+  [lifecycle-channel :- (schema/protocol async-prot/Channel)
+   shutdown-channel :- (schema/protocol async-prot/Channel)
    shutdown-reason-promise :- IDeref]
+  (log/debug "Initializing lifecycle worker loop.")
   (async/go (try
               (loop []
-                (let [task (async/<! lifecycle-chan)]
+                (let [[task chan] (async/alts!
+                                   [shutdown-channel lifecycle-channel]
+                                   :priority true)]
                   (schema/validate LifeCycleTask task)
                   (let [{:keys [type task-function]} task]
                     (condp #(contains? %1 %2) type
                       #{:shutdown}
                       (do
-                        (log/debug "Received shutdown command, lifecycle worker exiting.")
+                        (log/debug "Received shutdown command, shutting down services")
+                        (task-function)
+                        (log/debug "Service shutdown complete, exiting lifecycle worker loop")
+                        (async/close! shutdown-channel)
```

Comment:
Would it make sense to do the `async/close!` calls before the `task-function` one?  Since the `task-function` one could take a long time, I'm wondering if it would make sense to shut the incoming channels down right away since we already know once we've gotten into this code that we don't care to read anything else from them anyway.  Seems like that would allow any further putters onto the channel to be rejected more quickly?